### ISSUE TITLE
Add support for HostID-based node identities

### ISSUE THREAD

**Issue by rzetelskik** [scylladb/scylla-manager#3707]
As described in https://github.com/scylladb/scylla-manager/issues/1597, Scylla Manager currently collects node IPs from Scylla API and saves them as known_hosts.
Since we've started supporting hostID-based indentities ([https://github.com/scylladb/scylla-operator/issues/1311](https://github.com/scylladb/scylla-operator/issues/1311), [https://github.com/scylladb/scylla-operator/pull/1312](https://github.com/scylladb/scylla-operator/pull/1312)) with Scylla Operator, this has become an issue in cases when we're using ephemeral [PodIPs](https://operator.docs.scylladb.com/master/exposing.html#podip-type) for nodes' broadcast_addresses/broadcast_rpc_addresses.
For instance, this causes any rolling restarts to break connectivity with Scylla Manager, as it tries to reach the cluster using cached IPs, which are no longer in use by Scylla nodes. Bare in mind, though, that the issue is not limited to just rolling restarts, as any event causing the ephemeral IPs to be replaced will result in the same broken state.

A reproducible scenario would be to follow [the manual restore procedure](https://operator.docs.scylladb.com/stable/nodeoperations/restore.html) (a direct copy of the [restore procedure from Scylla Manager's docs](https://manager.docs.scylladb.com/branch-2.2/restore/index.html)). When using ephemeral IPs, performing a rolling restart after the schema has been restored will result in a broken state and the manager replying with errors similar to the following:
```
create restore target, units and views: create worker: get CQL cluster session: get client: discover cluster topology: dial tcp 10.85.0.72:10001: connect: no route to host
```

For context, Scylla Operator registers clusers with Scylla Manager using their DNS names.
I believe a change of an ephemeral IP behind it should not result in a broken state.

Afaic, this is a prerequisite for https://github.com/scylladb/scylla-operator/issues/1671, which stemmed from a customer issue https://github.com/scylladb/scylla-enterprise/issues/3746.

@karol-kokoszka @tzach can we prioritise this?

**Comment by karol-kokoszka** [comment]
I confirmed with @zimnx that we can release 3.2.6 with changes requested by operator. Including https://github.com/scylladb/scylla-manager/commit/c02571c2f8f44f76bd226ccc5c5b5a8e1fc5f081

Let me review all what's requested by operator team and label it with mielstone 3.2.6.

**Comment by karol-kokoszka** [comment]
`create restore target, units and views: create worker: get CQL cluster session: get client: discover cluster topology: dial tcp 10.85.0.72:10001: connect: no route to host`

.. apears on scylla client creation. We rely on this method to create HTTP client to work with scylla API (accessed through agent proxy)
https://github.com/scylladb/scylla-manager/blob/9a2c9fde8e9f2e11299eb18b0f50bf1884ce5cca/pkg/service/cluster/service.go#L101-L124

Client that is created, is load balancing calls to every nodes. So, if some of the IPs is changed, it may happen (and it happends) that the consecutive call to the API will fail as client will want to hit the node with non-existing IP. 
If it managed to `discoverHosts` then everything is it basically asks Scylla API to get list of node's IPs. And it updates this list to the DB.
`discoverHosts` is calling `/storage_service/host_id` that returns IP and HostID pair for every node that is part of the cluster.

To sum up:
- we cannot store HostIDs instead of IPs because these are used to create connection
- we cannot replace IPs with DNS names neither, as I see no way of getting DNS name by calling Scylla API (at least I haven't find endpoint in scylla swagger)

The solution I see, is to go through all the IP addresses that manager keeps in its cache and try to discoverHosts from these IPs. 
Error on current check, means that next IP in a row should be asked to discoverHosts.
If no node will respond with information about available IPs, then the cluster configuration in manager is broken and cluster must be added once again.

The propose solution doesn't solve other problem. The IP may be changes after the client is created.

**Comment by karol-kokoszka** [comment]
> I think the solution would be to at least store what was registered by the user during cluster registration, and fallback to it if knownHost stored in database are not reachable.

We are updating knownHosts every time when the client is being created.

The root cause of the problem described in this issue is a bit different.
We create client for scylla API that always keeps all the IPs and it's load balanced by the client.
Client creation is not returning any error if we pass non-existing IP in the configuration. Error will be returned when user wants to call some endpoint with the client and it will choose non-existing IP to execute the API call on it.
If the client is able to make the initial call (discoverHosts), then we are good to go, because it reads IPs from scylla configuration.
```
create restore target, units and views: create worker: get CQL cluster session: get client: discover cluster topology: dial tcp 10.85.0.72:10001: connect: no route to host
```
Error may appear on this initial call directly - discoverHosts and it happends if the non-existing IP will be chosen to handle the call.

I changed the approach to:
- interate over all known IPs
   - create the client with single IP
   - call to discover hosts 
   - if error then continue iteration
   - if not, then create the "final" client including all the IPs returned by ScyllaAPI

Besides, we are caching clients. I'm going to give possibility of disabling the cache. Otherwise, it may hit the client that is still in cache but keeps wrong IPs.

**Comment by karol-kokoszka** [comment]
> I think the solution would be to at least store what was registered by the user during cluster registration, and fallback to it if knownHost stored in database are not reachable.
We (Operator) register cluster using DNS name which always point to at least one one, even when all change their IP.

On the other hand, this proposal doesn't sound bad at all...
Operator could make use of it as you pass the DNS name on adding the cluster.
This DNS name could be used to discover IPs by calling scylla API using this DNS name.
The current approach could still be kept as a fallback to the above.

I'll add it to the PR.

**Comment by rzetelskik** [comment]
Thanks for looking into it, Karol.

> This DNS name could be used to discover IPs by calling scylla API using this DNS name.

This sounds reasonable. It's important, however, to also use it in case no of the known IPs are reachable, to cover the scenario of restarting the clusters with new IPs.

**Comment by karol-kokoszka** [comment]
> Thanks for looking into it, Karol.
> 
> > This DNS name could be used to discover IPs by calling scylla API using this DNS name.
> 
> This sounds reasonable. It's important, however, to also use it in case no of the known IPs are reachable, to cover the scenario of restarting the clusters with new IPs.

And it will be used when no IPs are reachable. I added it as a first check. Pr is ready.
